File: conn.go
//go:build js && wasm

package grpcwasm

import (
	"context"
	"syscall/js"

	"github.com/lesomnus/grpc-wasm/internal/jz"
	"google.golang.org/grpc"
	"google.golang.org/grpc/metadata"
	"google.golang.org/grpc/status"
)

type Conn struct {
	*grpc.ClientConn

	scope *jz.Scope
	ctx   context.Context
}

func (c *Conn) JsClose(this js.Value, args []js.Value) any {
	return c.scope.Promise(func() (js.Value, js.Value) {
		err := c.Close()
		if err != nil {
			return js.Undefined(), jz.ToError(err)
		}

		return js.Undefined(), js.Undefined()
	})
}

// JsInvoke send serialized data to the server with given method.
// It is resolved even if server responded with status.Code > 0 and "response" will be empty.
//
// Signature:
//
//	type Metadata = {
//		[key: string]: string[] | undefined
//	};
//	type RpcStatus = {
//		code: number
//		message: string
//	}
//	type RpcResult {
//		header: Metadata
//		trailer: Metadata
//		response: Uint8Array
//		status: RpcStatus
//	};
//	function(method: string, req: Uint8Array, option: {meta?: Metadata}): Promise<RpcResult>;
func (c *Conn) JsInvoke(this js.Value, args []js.Value) any {
	return c.scope.Promise(func() (js.Value, js.Value) {
		if len(args) != 3 {
			return js.Undefined(), jz.Error("expects 3 arguments: method, serialized message, and option")
		}

		method := args[0].String()
		req := args[1]
		opt := args[2]

		size := req.Get("byteLength").Int()
		data := make([]byte, size)
		js.CopyBytesToGo(data, req)

		ctx := c.ctx

		meta := metadata.MD{}
		if v := opt.Get("meta"); v.Type() == js.TypeObject {
			metaToGo(meta, v)
		}
		if meta.Len() > 0 {
			ctx = metadata.NewOutgoingContext(ctx, meta)
		}

		header := metadata.MD{}
		trailer := metadata.MD{}
		opts := []grpc.CallOption{
			grpc.Header(&header),
			grpc.Trailer(&trailer),
		}

		var (
			out []byte
			st  status.Status
		)
		if err := c.ClientConn.Invoke(ctx, method, data, &out, opts...); err != nil {
			s, ok := status.FromError(err)
			if !ok {
				return js.Undefined(), jz.ToError(err)
			}
			if s != nil {
				st = *s
			}
		}

		js_out := js.Global().Get("Uint8Array").New(len(out))
		js.CopyBytesToJS(js_out, out)

		return js.ValueOf(map[string]any{
			"header":   metaToJs(header),
			"trailer":  metaToJs(trailer),
			"response": js_out,
			"status":   statusToJs(&st),
		}), js.Undefined()
	})
}

// Signature:
//
//	type StreamResult =
//		| {
//			response: Uint8Array;
//		}
//		| {
//			trailer: Metadata;
//			status: RpcStatus;
//		};
//	type Stream = {
//		header: ()=>Promise<Metadata>
//		close: ()=>Promise<void>
//		close_send: ()=>Promise<void>
//		send: (Uint8Array)=>Promise<void>
//		recv: ()=>Promise<StreamResult>
//	}
//	function(method: string, option: {meta?: Metadata}): Promise<Stream>;
func (c *Conn) jsOpenStream(desc *grpc.StreamDesc, _ js.Value, args []js.Value) any {
	return c.scope.Promise(func() (js.Value, js.Value) {
		if len(args) != 2 {
			return js.Undefined(), jz.Error("expects 2 arguments: method, and option")
		}

		method := args[0].String()
		opt := args[1]

		ctx := c.ctx
		meta := metadata.MD{}
		if v := opt.Get("meta"); v.Type() == js.TypeObject {
			metaToGo(meta, v)
		}
		if meta.Len() > 0 {
			ctx = metadata.NewOutgoingContext(ctx, meta)
		}

		stream, err := NewStream(ctx, c, desc, method)
		if err != nil {
			return js.Undefined(), jz.ToError(err)
		}

		return stream.ToJs(), js.Undefined()
	})
}

func (c *Conn) JsOpenServerStream(this js.Value, args []js.Value) any {
	return c.jsOpenStream(&grpc.StreamDesc{ServerStreams: true}, this, args)
}

func (c *Conn) JsOpenClientStream(this js.Value, args []js.Value) any {
	return c.jsOpenStream(&grpc.StreamDesc{ClientStreams: true}, this, args)
}

func (c *Conn) JsOpenBidiStream(this js.Value, args []js.Value) any {
	return c.jsOpenStream(&grpc.StreamDesc{ServerStreams: true, ClientStreams: true}, this, args)
}

func (c Conn) ToJs() js.Value {
	return js.ValueOf(map[string]any{
		"close":  c.scope.FuncOf(c.JsClose),
		"invoke": c.scope.FuncOf(c.JsInvoke),

		"open_server_stream": c.scope.FuncOf(c.JsOpenServerStream),
		"open_client_stream": c.scope.FuncOf(c.JsOpenClientStream),
		"open_bidi_stream":   c.scope.FuncOf(c.JsOpenBidiStream),
	})
}

File: src/conn.ts
import {
	BidiStream,
	type BidiStreamingClient,
	type ClientStreamingClient,
	type ServerStreamingClient,
} from "./stream";
import type { CallOption, InvokeOption, RpcResult } from "./types";
import type { BridgeWorker, ConnId } from "./worker";

export interface Conn {
	close(): Promise<void>;
	invoke(method: string, req: Uint8Array, option: InvokeOption): Promise<RpcResult>;
	open_server_stream(
		method: string,
		req: Uint8Array,
		option: CallOption,
	): Promise<ServerStreamingClient>;
	open_client_stream(method: string, option: CallOption): Promise<ClientStreamingClient>;
	open_bidi_stream(method: string, option: CallOption): Promise<BidiStreamingClient>;
}

export class ClientConn implements Conn {
	private close_work: Promise<void> | undefined;

	constructor(
		private worker: BridgeWorker,
		private id: ConnId,
	) {}

	close(): Promise<void> {
		if (this.close_work) {
			return this.close_work;
		}

		this.close_work = this.worker.close(this.id);
		return this.close_work;
	}

	async invoke(method: string, req: Uint8Array, option: InvokeOption): Promise<RpcResult> {
		this.throwIfClosed();

		const id = await this.worker.invoke(this.id, method, req, {
			meta: option.meta,
		});

		// `recv` must be called before `cancel`
		// since `cancel` deletes the handle.
		const result = this.worker.recv(id);

		if (option.signal) {
			if (option.signal.aborted) {
				this.worker.cancel(id);
			} else {
				option.signal.addEventListener("abort", () => {
					this.worker.cancel(id);
				});
			}
		}
		return result;
	}

	async open_server_stream(
		method: string,
		req: Uint8Array,
		option: CallOption,
	): Promise<ServerStreamingClient> {
		this.throwIfClosed();

		const stream_id = await this.worker.open_server_stream(this.id, method, option);
		const stream = new BidiStream(this.worker, stream_id);
		await stream.send(req);
		await stream.close_send();
		return stream;
	}

	async open_client_stream(method: string, option: CallOption): Promise<ClientStreamingClient> {
		this.throwIfClosed();

		const stream_id = await this.worker.open_client_stream(this.id, method, option);
		return new BidiStream(this.worker, stream_id);
	}

	async open_bidi_stream(method: string, option: CallOption): Promise<BidiStreamingClient> {
		this.throwIfClosed();

		const stream_id = await this.worker.open_bidi_stream(this.id, method, option);
		return new BidiStream(this.worker, stream_id);
	}

	private throwIfClosed() {
		if (this.close_work) {
			throw new Error("closed");
		}
	}
}
